jinad/api/endpoints/logs.py

```diff
@@ -1,42 +1,55 @@
 import uuid
+import time
 import asyncio
+from pathlib import Path
+from typing import Optional
 
-from fastapi import APIRouter
-from sse_starlette.sse import EventSourceResponse
+from jina.logging import JinaLogger
+from fastapi import APIRouter, WebSocket, WebSocketDisconnect
 
 from config import log_config
+from excepts import HTTPException
 
+logger = JinaLogger(context='👻 LOGS')
 router = APIRouter()
 
 
-@router.get(
-    path='/log/{log_id}',
-    summary='Stream logs from folder where fluentd stores',
+@router.websocket(
+    path='/wslog/{log_id}'
 )
-async def _log(
-        log_id: uuid.UUID
+async def _websocket_logs(
+    websocket: WebSocket,
+    log_id: uuid.UUID,
 ):
-    # TODO: How to ever stop this iteration?
     """
-    This endpoint is intended to be used from external clients (like dashboard) or from jina Flows or Pods themselves to
-    print to log themselves this streamed data.
+    # TODO: Swagger doesn't support websocket based docs
+    Websocket endpoint to stream logs from fluentd file
 
-    :param log_id: The log identifier `flow_id`, `pod_id` or `pea_id` from which to stream logs
-    :return:
+    ```
+    `log_id`: uuid of the flow/pod/pea
+    ```
     """
-    file_path = f'{log_config.LOG_BASE_PATH}/{log_id}/log.log'
-
-    async def _log_stream(file):
-        # fluentd creates files under this path with some tag based on day, so as temp solution,
-        # just get the first file matching this patter once it appears
-        with open(file) as fp:
-            fp.seek(0, 2)
-            while True:
-                readline = fp.readline()
-                line = readline.strip()
-                if line:
-                    yield f'data: {line}\n\n'
-                else:
-                    await asyncio.sleep(0.1)
-
-    return EventSourceResponse(_log_stream(file_path), media_type='text/event-stream')
+    file_path = log_config.PATH % log_id
+    if not Path(file_path).is_file():
+        raise HTTPException(status_code=404,
+                            detail=f'No logs found')
+    await websocket.accept()
+    client_host = websocket.client.host
+    client_port = websocket.client.port
+    logger.info(f'Client {client_host}:{client_port} got connected!')
+    try:
+        while True:
+            data = await websocket.receive_json()
+            d = {}
+            line_num_from = int(data.get('from', 0))
+            line_num_to = int(data.get('to', line_num_from + 5))
+            with open(file_path) as fp:
+                for line_num, line in enumerate(fp):
```

Comment:
will this follow as the file is being filled?

Reply:
Yes. with every websocket request, we'll start reading the file.

Reply:
I do not understand how it works actually. How does this follow the file continuosly growing?